conn_linux.go

```diff
@@ -344,32 +346,50 @@ func newSysSocket(lockThread bool) *sysSocket {
 
 	funcC := make(chan func())
 	doneC := make(chan bool)
+	errC := make(chan error)
 
 	go func() {
 		// It is important to lock this goroutine to its OS thread for the duration
 		// of the netlink socket being used, or else the kernel may end up routing
 		// messages to the wrong places.
 		// See: http://lists.infradead.org/pipermail/libnl/2017-February/002293.html.
 		//
-		// But since this is very experimental, we'll leave it as a configurable at
-		// this point.
-		if lockThread {
-			// The intent is to never unlock the OS thread, so that the thread
-			// will terminate when the goroutine exits starting in Go 1.10:
-			// https://go-review.googlesource.com/c/go/+/46038.
-			//
-			// However, due to recent instability and a potential bad interaction
-			// with the Go runtime for threads which are not unlocked, we have
-			// elected to temporarily unlock the thread:
-			// https://github.com/golang/go/issues/25128#issuecomment-410764489.
-			//
-			// If we ever allow a Conn to set its own network namespace, we must
-			// either ensure that the namespace is restored on exit here or that
-			// the thread is properly terminated at some point in the future.
-			runtime.LockOSThread()
-			defer runtime.UnlockOSThread()
+		// The intent is to never unlock the OS thread, so that the thread
+		// will terminate when the goroutine exits starting in Go 1.10:
+		// https://go-review.googlesource.com/c/go/+/46038.
+		//
+		// However, due to recent instability and a potential bad interaction
+		// with the Go runtime for threads which are not unlocked, we have
+		// elected to temporarily unlock the thread when the goroutine terminates:
+		// https://github.com/golang/go/issues/25128#issuecomment-410764489.
+
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
+
+		// The user requested the Conn to operate in a non-default network namespace.
+		if config.NetNS != 0 {
+
+			// Get the current namespace of the thread the goroutine is locked to.
+			origNetNS, err := getThreadNetNS()
+			if err != nil {
+				errC <- err
```

Comment:
On error need to return.

Should probably move defer on line 393 up to roughly line 368.